File: Binary Tree/513.找树左下角的值.cpp
#include <queue>

using namespace std;
/*
 * @lc app=leetcode.cn id=513 lang=cpp
 *
 * [513] 找树左下角的值
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int findBottomLeftValue(TreeNode *root)
    {
        int result;
        queue<TreeNode *> qu;

        if (root != nullptr)
        {
            qu.push(root);
        }

        while (!qu.empty())
        {
            int size = qu.size();
            bool notFind = true;

            for (int i = 0; i < size; i++)
            {
                TreeNode *temp = qu.front();
                qu.pop();

                // 找到当前层级的最左边子节点
                // 由于左右节点是成对放入队列的，所以左节点在队列中的下标一定是偶数0,2,4
                if (notFind && temp != nullptr)
                {
                    result = temp->val;
                    notFind = false;
                }
                if (temp != nullptr)
                {
                    qu.push(temp->left);
                    qu.push(temp->right);
                }
            }
        }

        return result;
    }
};
// @lc code=end

File: Binary Tree/二叉树的属性/513.找树左下角的值.ts
/*
 * @lc app=leetcode.cn id=513 lang=typescript
 *
 * [513] 找树左下角的值
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function findBottomLeftValue(root: TreeNode | null): number {
  let result = 0
  let maxLevel = 0

  // 寻找最左子节点递归函数‘
  const findLastLeft = (level: number, node: TreeNode | null): void => {
    if (!node) {
      return
    }

    // 符合条件
    if (!node.left && !node.right) {
      if (level > maxLevel) {
        result = node.val
        maxLevel = level
      }
      return
    }

    findLastLeft(level + 1, node.left)
    findLastLeft(level + 1, node.right)
  }

  findLastLeft(1, root)

  return result
};
// @lc code=end

